samples/go/ipfs-chat/daemon.go

```diff
@@ -20,121 +26,186 @@ import (
 		}
-		if a.Equals(destDS.HeadRef()) {
+		if a.Equals(localDS.HeadRef()) {
 			dbg.Debug("fast-forward to source commit")
-			destDS, err = destDB.SetHead(destDS, sourceRef)
-			didChange(destDS)
+			localDS, err = localDB.SetHead(localDS, sourceRef)
+			didChange(localDS)
 			continue
 		}
 
-		left := destDS.HeadValue()
+        dbg.Debug("We have mergeable commit")
+		left := localDS.HeadValue()
 		right := sourceCommit.(types.Struct).Get("value")
-		parent := a.TargetValue(destDB).(types.Struct).Get("value")
+		parent := a.TargetValue(localDB).(types.Struct).Get("value")
 
-		merged, err := merge.ThreeWay(left, right, parent, destDB, nil, nil)
+        dbg.Debug("Starting three-way commit")
+		merged, err := merge.ThreeWay(left, right, parent, localDB, nil, nil)
 		if err != nil {
 			dbg.Debug("could not merge received data: " + err.Error())
 			continue
 		}
 
-		destDS, err = destDB.SetHead(destDS, destDB.WriteValue(datas.NewCommit(merged, types.NewSet(destDB, destDS.HeadRef(), sourceRef), types.EmptyStruct)))
+        dbg.Debug("setting new datasetHead on localDB")
+        localDS, err = localDB.SetHead(localDS, localDB.WriteValue(datas.NewCommit(merged, types.NewSet(localDB, localDS.HeadRef(), sourceRef), types.EmptyStruct)))
 		if err != nil {
-			dbg.Debug("call failed to SetHead on destDB, err: %s", err)
+			dbg.Debug("call failed to SetHead on localDB, err: %s", err)
 		}
-		didChange(destDS)
+        dbg.Debug("merged commit, new dataset head is: %ss", localDS.HeadRef().TargetHash())
+		didChange(localDS)
 	}
 }
 
-func pullCommits(h hash.Hash, sourceDB, destDB datas.Database) {
-	dbg.Debug("pullCommits, h: %s", h)
-
-	// read
-	v := destDB.ReadValue(h)
-	if v != nil {
-		dbg.Debug("pullCommits, found h: %s, commit:", h, types.EncodedValueMaxLines(v, 10))
+// One of the problem that this daemon currently has is that it hangs which
+// effectively stops are 'replicate' loop. I can't figure out why that is
+// happening.
+func pullCommits(h hash.Hash, netDB, localDB datas.Database, level int) {
+    // This is so we can easily tell in the log that we are not blocked in this
+    // function
+    defer func() {
+        if level == 0 {
+            dbg.Debug("EXITING PULL-COMMITS!!!")
+        }
+    }()
+    
+    // This is an optimization that can lead us to a bad place, if for some
+    // reason we find a commit in the localDB that doesn't have all of it's
+    // parents then something is wrong.
+	dbg.Debug("pullCommits, checking in local, h: %s", h)
+	if localDB.ReadValue(h) != nil {
+		dbg.Debug("pullCommits, found local h: %s", h)
 		return
-	}
-	v = sourceDB.ReadValue(h)
-	types.EncodedValue(v)
-	dbg.Debug("pullCommits, read h: %s, commit:", h, types.EncodedValueMaxLines(v, 10))
+    }
+    
+    dbg.Debug("pullCommits, not found in local, reading from net h: %s", h)
+    v := netDB.ReadValue(h)
+    d.Chk.NotNil(v)
+    
+    dbg.Debug("pullCommits, encoding value from net, h: %s", h)
+    s1 := types.EncodedValue(v)
```

Comment:
EncodedValue is pretty heavy (it generates a huge string). You could accomplish same thing by calling `value.WalkRefs() -> ref.TargetValue() -> value.WalkRefs()` recursively.

Reply:
I also continue to feel like you could generalize this and optimize it at the same time by using pinned-ness rather than present-in-local-db-ness.

So basically this would recurse an arbitrary noms ref (typically starting at a commit) reading and pinning all its child refs. If it ever finds a ref that is already pinned it can stop.

Then I think you might not even need `ipfs-local` at all?